### PR TITLE
gh-103987: fix crash in mmap module

### PR DESCRIPTION
**Pull request by Agent-Hellboy**
* Issue: gh-103987

### CODE REVIEW

**Comment by Agent-Hellboy** [comment]
I am closing the mmap object still the test is failing

File: Lib/test/test_mmap.py [hunk 1]
```diff
@@ -264,6 +264,26 @@ def test_bad_file_desc(self):
         # Try opening a bad file descriptor...
         self.assertRaises(OSError, mmap.mmap, -2, 4096)
 
+    def test_if_crash(self): # test for issue gh-103987
+        with open("TESTFN", "wb+") as f:
```
Comment (Eclips4) [inline Lib/test/test_mmap.py]:
```suggestion
        with open(TESTFN, "wb+") as f:
```

Comment (Eclips4) [reply Lib/test/test_mmap.py]: Test doesn't failed. This test affects on execution environment, so, CI/CD fails =)

File: Lib/test/test_mmap.py [hunk 2]
```diff
@@ -264,6 +264,26 @@ def test_bad_file_desc(self):
         # Try opening a bad file descriptor...
         self.assertRaises(OSError, mmap.mmap, -2, 4096)
 
+    def test_if_crash(self): # test for issue gh-103987
```
Comment (Eclips4) [inline Lib/test/test_mmap.py]:
```suggestion
    def test_if_crash(self):
        # See gh-103987
```

Comment (Eclips4) [reply Lib/test/test_mmap.py]: Also, I'd like to see more meaningful name for this test-case, but I have no ideas :(
Maybe, `test_unexpected_mmap_close`?

Comment (Agent-Hellboy) [reply Lib/test/test_mmap.py]: sure

**Comment by Eclips4** [comment]
Let's wait CI/CD results. Also, can you add a few lines about this in `NEWS` entry?

**Comment by Eclips4** [comment]
Great work @Agent-Hellboy !
Let's see what core devs will say.

**Comment by cfbolz** [comment]
still some cases not covered properly. Now that you moved the `CHECK_VALID(NULL);` into the if-branch, you don't check whether the mmap is closed any more if the index is a slice. try adding a test like:

```python
m = mmap(...)
m.close()
m[0:5] # should complain about a closed file
```

also you should add a test that uses the weird `X` class as part of a slice

```python
m[X():5]
```

and you need to do the same thing again for assignment, both with `X` as an index and a slice:

```python
m[X()] = 1
...
m[X():5] = b'abcde'
```

or something like that.

Also there needs to be a test for the special case I mentioned at the end of the issue:

```python
m = mmap(...)
m.close()
with self.assertRaises(ValueError):
    m["abc"] # not TypeError!
```

**Comment by Agent-Hellboy** [comment]
hi @cfbolz, Now I am checking at the start as well which denies thrwoing TypeError for closed mmap

**Comment by cfbolz** [comment]
now you need to do the same things for item assignments still.

File: Modules/mmapmodule.c [hunk 3]
```diff
@@ -936,6 +936,7 @@ mmap_subscript(mmap_object *self, PyObject *item)
     CHECK_VALID(NULL);
     if (PyIndex_Check(item)) {
         Py_ssize_t i = PyNumber_AsSsize_t(item, PyExc_IndexError);
+        CHECK_VALID(NULL);
```
Comment (sunmy2019) [inline Modules/mmapmodule.c]: I think it's better to add it right before 
```c
        return PyLong_FromLong(Py_CHARMASK(self->data[i]));
```
which maintains the best compatibility.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

**Comment by sunmy2019** [comment]
We should `CHECK_VALID(NULL);` before each `self->data` access, unless we have strong proof that our handle is not closed.

**Comment by sunmy2019** [comment]
@Agent-Hellboy I am not suggesting this.

I am saying you should add these

```c
    else if (PySlice_Check(item)) {
        Py_ssize_t start, stop, step, slicelen;
        if (PySlice_Unpack(item, &start, &stop, &step) < 0) {
            return NULL;
        }
        CHECK_VALID(NULL); /////////////////////////////////////////////////////////////////
        slicelen = PySlice_AdjustIndices(self->size, &start, &stop, step);
        if (slicelen <= 0)
            return PyBytes_FromStringAndSize("", 0);
        else if (step == 1)
            return PyBytes_FromStringAndSize(self->data + start,
                                              slicelen);
        else {
            char *result_buf = (char *)PyMem_Malloc(slicelen);
            size_t cur;
            Py_ssize_t i;
            PyObject *result;
            if (result_buf == NULL)
                return PyErr_NoMemory();
            CHECK_VALID(NULL); /////////////////////////////////////////////////////////////////

            for (cur = start, i = 0; i < slicelen;
                 cur += step, i++) {
                result_buf[i] = self->data[cur];
            }
            result = PyBytes_FromStringAndSize(result_buf,
                                                slicelen);
            PyMem_Free(result_buf);
            return result;
        }
    }
    else {
        PyErr_SetString(PyExc_TypeError,
                        "mmap indices must be integers");
        return NULL;
    }
}
```

**Comment by Agent-Hellboy** [comment]
Hi @sunmy2019, please share with me a scenario where it's bypassing the first  `CHECK_VALID(NULL);` at the start for a slice

**Comment by sunmy2019** [comment]
> Hi @sunmy2019, please share with me a scenario where it's bypassing the first `CHECK_VALID(NULL);` at the start for a slice

`m[X():5]`

**Comment by Agent-Hellboy** [comment]
Currently, it's throwing Value error for the same, i have added a test for the same

File: Lib/test/test_mmap.py [hunk 4]
```diff
@@ -264,6 +264,57 @@ def test_bad_file_desc(self):
...
+                    m.close()
+                    return 1
+
+            m = mmap.mmap(f.fileno(), n, access=mmap.ACCESS_READ)
+
+        self.assertEqual(m[1], 97)
+        with self.assertRaises(ValueError):
+            m[X()] # should not crash
+
+        with self.assertRaises(ValueError):
```
Comment (sunmy2019) [inline Lib/test/test_mmap.py]: You should create a new `m` instead of using the closed `m` used before.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

**Comment by sunmy2019** [comment]
> Currently, it's throwing expected Value error

This is wrong. And you are not creating a new `m`, which covered your mistake.

**Comment by Agent-Hellboy** [comment]
> > Currently, it's throwing expected Value error
> 
> This is wrong. And you are not creating a new `m`, which covered your mistake.

yes, silly me, adding checks in PySlice_Check

**Comment by Agent-Hellboy** [comment]
I still need to add checks while mmap is opened in witemode , will add test for the same as well
it's still crashing for 
```
m[X()] = 1
...
m[X():5] = b'abcde'
```

**Comment by Agent-Hellboy** [comment]
HI @sunmy2019, I don't have experience with c, how can I check the core dump generated during a crash to explore the call stack using `gcc dump_file core`

**Comment by sunmy2019** [comment]
> is there a flag for make to compile in debug mode

`./configure --with-pydebug`

You can use `gdb python`, then `r ARG1 ARG2 ...` , it will run `python ARG1 ARG2 ...` in gdb.

But there is a simpler approach here. Search plain text `->data` in `Modules/mmapmodule.c`

**Comment by sunmy2019** [comment]
You should add checks in `mmap_ass_subscript`

**Comment by JelleZijlstra** [comment]
I think there's another one in `mmap_resize_method`: we do CHECK_VALID, then PyArg_ParseTuple, then we access self->fd. But I think `PyArg_ParseTuple` can trigger GC, so it could invoke Python code and invalidate the mmap.

**Comment by Agent-Hellboy** [comment]
HI, @sunmy2019 and @JelleZijlstra  do I need to put a check before every self->data access in that function or do I need to find a place where memory move can happen in mmap because gc cleans that memory? I have very less experience using c and this is the first time I am exploring the CPython code base. please explain in layman's terms to me

**Comment by sunmy2019** [comment]
> do I need to find a place where memory move can happen in mmap because gc cleans that memory

I cannot understand this



> do I need to put a check before every self->data access

In theory, yes. Most cases are already guarded. This issue mainly want to fix the scenarios that
1. check_valid
2. execute python codes
3. read without check_valid

Since python codes may invalidate it.

**Comment by Agent-Hellboy** [comment]
>I cannot understand this
Hi, I guess during resize gc will clear the memory for optimization or something else it gets new memory, and the file object will try to access that memory which is not there as gc has cleared it and file object(fd) become invalid

**Comment by Agent-Hellboy** [comment]
HI, @JelleZijlstra I am not getting your point, using `resize()` on a closed object should crash somehow.

**Comment by JelleZijlstra** [comment]
> HI, @JelleZijlstra I am not getting your point, using `resize()` on a closed object should crash somehow.
> 
> did you mean `PyArg_ParseTuple` can trigger Python code to close the mmap in between?

Yes.

**Comment by sunmy2019** [comment]
Seems no update then.

I will make it work tonight.

**Comment by Agent-Hellboy** [comment]
>Seems there is no update.

sorry, I didn't find the code for Unittest. I should have updated it here.

File: Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix ``mmap_subscript`` in mmapmodule.c to avoid segmentation fault
```
Comment (sunmy2019) [inline Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst]:
```suggestion
add more `CHECK_VALID`s in `mmapmodule.c` to avoid the file being invalidated by Python code.
```

File: Lib/test/test_mmap.py [hunk 6]
```diff
@@ -264,6 +264,111 @@ def test_bad_file_desc(self):
...
+            data = b'aabaac\x00deef\x00\x00aa\x00'
+            n = len(data)
+            f.write(data)
+            f.flush()
+
+            m = mmap.mmap(f.fileno(), n, access=mmap.ACCESS_READ)
+
+        m.close()
+        with self.assertRaises(ValueError):
+            m["abc"]
```
Comment (sunmy2019) [inline Lib/test/test_mmap.py]: This is not **unexpected** mmap close. Remove this test, or move to other places.

Also, add one more test `m[X():5:1]`.

File: Lib/test/test_mmap.py [hunk 7]
```diff
@@ -264,6 +264,111 @@ def test_bad_file_desc(self):
...
+            data = b'aabaac\x00deef\x00\x00aa\x00'
+            n = len(data)
+            f.write(data)
+            f.flush()
+
+            m = mmap.mmap(f.fileno(), n, access=mmap.ACCESS_READ)
+
+        m.close()
+        with self.assertRaises(ValueError):
+            m[0:5] # should not crash
```
Comment (sunmy2019) [inline Lib/test/test_mmap.py]: This is not **unexpected** mmap close. Remove this test, or move to other places.

**Review by sunmy2019 — CHANGES_REQUESTED** [review]

File: Lib/test/test_mmap.py (hunk 6 above)
Comment (Agent-Hellboy) [reply Lib/test/test_mmap.py]: removed

File: Lib/test/test_mmap.py (hunk 7 above)
Comment (Agent-Hellboy) [reply Lib/test/test_mmap.py]: removed

File: Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+add more `CHECK_VALID`s in `mmapmodule.c` to avoid the file being invalidated by Python code.
```
Comment (sunmy2019) [inline Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst]:
```suggestion
add more ``CHECK_VALID``s in ``mmapmodule.c`` to avoid the file being invalidated by Python code.
```

File: Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst [hunk 9]
```diff
@@ -0,0 +1 @@
+add more ``CHECK_VALID``s in ``mmapmodule.c`` to avoid the file being invalidated by Python code.
```
Comment (sunmy2019) [inline Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst]:
```suggestion
add more ``CHECK_VALID`` s in ``mmapmodule.c`` to avoid the file being invalidated by Python code.
```

File: Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst [hunk 10]
```diff
@@ -0,0 +1 @@
+add more ``CHECK_VALID`` s in ``mmapmodule.c`` to avoid the file being invalidated by Python code.
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2023-04-29-18-23-16.gh-issue-103987.sRgALL.rst]: Focus on what this means to users, not on the mechanism used to achieve it.
```suggestion
In :mod:`mmap`, fix several bugs that could lead to access to memory-mapped files after
they have been invalidated.
```

**Comment by sunmy2019** [comment]
> then we access self->fd

I think checking `self->fd != -1` is enough here.

**Comment by sunmy2019** [comment]
I think it's now ready for review. @JelleZijlstra @cfbolz

Sorry for the delay!

File: Lib/test/test_mmap.py [hunk 11]
```diff
@@ -887,6 +886,92 @@ def test_resize_succeeds_with_error_for_second_named_mapping(self):
         self.assertEqual(m1[:data_length], data)
         self.assertEqual(m2[:data_length], data)
 
+    def test_mmap_closed_by_int_scenarios(self):
+        """
+        gh-103987: Test that mmap objects raise ValueError
+                for accidentally closed mmap files
```
Comment (JelleZijlstra) [inline Lib/test/test_mmap.py]:
```suggestion
                for closed mmap files
```
Pretty sure we're doing it on purpose

File: Modules/mmapmodule.c [hunk 12]
```diff
@@ -968,6 +976,7 @@ mmap_subscript(mmap_object *self, PyObject *item)
 
             if (result_buf == NULL)
                 return PyErr_NoMemory();
+
```
Comment (JelleZijlstra) [inline Modules/mmapmodule.c:979]: Could the buffer have been invalidated here as a side effect of the PyMem_Malloc call above triggering GC? I seem to recall that's possible, but not 100% sure.

Comment (sunmy2019) [reply Modules/mmapmodule.c:979]: I cannot find any GC-related code in `Object/obmalloc.c`
